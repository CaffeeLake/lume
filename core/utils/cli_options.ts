import { parseArgs } from "../../deps/cli.ts";
import type { DeepPartial } from "./object.ts";
import type { SiteOptions } from "../site.ts";

export function getOptionsFromCli(
  options: DeepPartial<SiteOptions>,
): DeepPartial<SiteOptions> {
  const cli = parseArgs(Deno.args, {
    string: ["src", "dest", "location", "port"],
    boolean: ["serve", "open"],
    alias: { dev: "d", serve: "s", port: "p", open: "o" },
    ["--"]: true,
  });

  if (cli.src) {
    options.src = cli.src;
  }

  if (cli.dest) {
    options.dest = cli.dest;
  }

  if (cli.location) {
    options.location = new URL(cli.location);
  } else if (cli.serve || cli._[0] === "cms") {
    options.location = new URL(`http://localhost:${cli.port || 3000}/`);
  }

  if (cli.port) {
    (options.server ||= {}).port = parseInt(cli.port);

    if (options.location) {
      options.location.port = cli.port;
    }
  }

  if (cli.open) {
    (options.server ||= {}).open = cli.open;
  }

  return options;
}
