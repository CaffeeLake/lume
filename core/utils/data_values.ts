import type { Data } from "../file.ts";

/**
 * Get the value of a page data
 * For example, if the value is "=title", it returns the value of the page data "title"
 * If the value is "$.title", it will return the value of the element with the selector ".title"
 */
export function getDataValue(data: Partial<Data>, value?: unknown) {
  // Get the value from the page data
  if (typeof value === "string") {
    if (value.startsWith("=")) {
      const key = value.slice(1);

      if (!key.includes(".")) {
        return data[key];
      }

      const keys = key.split(".");
      let val = data;
      for (const key of keys) {
        val = val?.[key];
      }
      return val;
    }

    if (value.startsWith("$")) {
      const document = data.page?.document;
      const query = value.slice(1);
      const element = document?.querySelector(query);
      return element?.innerHTML;
    }
  }

  if (typeof value === "function") {
    return value(data);
  }

  return value;
}
