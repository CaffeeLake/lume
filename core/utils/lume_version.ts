import { cyan, green, red } from "../../deps/colors.ts";

/** Return the latest stable version from the deno.land/x repository */
export async function getLatestVersion(): Promise<string> {
  const response = await fetch("https://cdn.deno.land/lume/meta/versions.json");
  const versions = await response.json();
  return versions.latest;
}

/** Return the hash of the latest commit from the GitHub repository */
export async function getLatestDevelopmentVersion(): Promise<string> {
  const response = await fetch(
    `https://api.github.com/repos/lumeland/lume/commits/main`,
  );
  const commits = await response.json();
  return commits.sha;
}

/** Return the current installed version */
export function getCurrentVersion(
  url = new URL(import.meta.resolve("../")),
): string {
  const { pathname } = url;
  return pathname.match(/@([^/]+)/)?.[1] ?? `local (${pathname})`;
}

/** Check the compatibility with the current Deno version */
export function checkDenoSupport(): void {
  const minimum = "1.38.0";
  const current = Deno.version.deno;

  if (current < minimum) {
    console.log("----------------------------------------");
    console.error(red("Your Deno version is not compatible with Lume"));
    console.log(`Lume needs Deno ${green(minimum)} or greater`);
    console.log(`Your current version is ${red(current)}`);
    console.log(`Run ${cyan("deno upgrade")} and try again`);
    console.log("----------------------------------------");
    Deno.exit(1);
  }
}
