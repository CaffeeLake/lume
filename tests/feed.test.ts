import { assertSiteSnapshot, build, getSite } from "./utils.ts";
import feed from "../plugins/feed.ts";

Deno.test("RSS plugin", async (t) => {
  const site = getSite({
    src: "normal",
    location: new URL("https://example.com/"),
  });

  site.use(
    feed({
      output: ["feed.json", "feed.rss"],
      info: {
        published: new Date("2020-01-01"),
        generator: "https://lume.land",
      },
      items: {
        title: (data) => data.title?.toUpperCase(),
        updated: "=date",
        image: "=metas.image",
      },
    }),
  );
  site.ignore("static.yml");

  // build page
  await build(site);

  // compare to snapshot
  await assertSiteSnapshot(t, site);
});
