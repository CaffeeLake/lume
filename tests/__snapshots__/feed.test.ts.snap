export const snapshot = {};

snapshot[`RSS plugin 1`] = `
{
  formats: [
    {
      engines: 0,
      ext: ".page.toml",
      loader: [AsyncFunction: toml],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.ts",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.js",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.jsonc",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.json",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".json",
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".jsonc",
      loader: [AsyncFunction: json],
    },
    {
      engines: 1,
      ext: ".md",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".markdown",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
      loader: [AsyncFunction: module],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".vento",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".vto",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: 0,
      ext: ".toml",
      loader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yaml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
  ],
  src: [
    "/",
    "/404.md",
    "/_data.yml",
    "/favicon.png",
    "/images",
    "/images/avatar.jpg",
    "/page5.yaml",
    "/pages",
    "/pages/2020-06-21_page2.page.json",
    "/pages/2021-01-02-18-32_page4.page.ts",
    "/pages/_data",
    "/pages/_data.yml",
    "/pages/_data/colors.yml",
    "/pages/_data/documents.ts",
    "/pages/_data/drinks.js",
    "/pages/_data/names.json",
    "/pages/ghost",
    "/pages/ghost/2021-12-29-page6.md",
    "/pages/ghost/_data.yml",
    "/pages/page1.md",
    "/pages/page3.page.js",
    "/pages/subpage",
    "/pages/subpage/_data.yml",
    "/pages/subpage/page7.page.js",
    "/static.yml",
    "/styles.css",
  ],
}
`;

snapshot[`RSS plugin 2`] = `[]`;

snapshot[`RSS plugin 3`] = `
[
  {
    content: '{"version":"https://jsonfeed.org/version/1","title":"My RSS Feed","home_page_url":"https://example.com/","feed_url":"https://example.com/feed.json","description":"","items":[{"id":"https://example.com/pages/new-name/page7/","url":"https://example.com/pages/new-name/page7/","content_html":"Content of Page 7","date_published":"Sun, 02 Jan 2022 00:00:00 GMT","date_modified":"Sun, 02 Jan 2022 00:00:00 GMT"},{"id":"https://example.com/pages/page6/","url":"https://example.com/pages/page6/","title":"PAGE 6","content_html":"<p>Content of Page 6</p>\\\\n","date_published":"Sat, 01 Jan 2022 00:00:00 GMT","date_modified":"Sat, 01 Jan 2022 00:00:00 GMT"},{"id":"https://example.com/pages/page4/","url":"https://example.com/pages/page4/","title":"PAGE 4","content_html":"Content of Page 4 in Overrided site name, from the file /pages/2021-01-02-18-32_page4.page.ts","date_published":"Sat, 02 Jan 2021 18:32:00 GMT","date_modified":"Sat, 02 Jan 2021 18:32:00 GMT"},{"id":"https://example.com/overrided-page2/","url":"https://example.com/overrided-page2/","title":"PAGE 2","content_html":"Content of Page 2","date_published":"Sun, 21 Jun 2020 00:00:00 GMT","date_modified":"Sun, 21 Jun 2020 00:00:00 GMT","image":"https://example.com/images/avatar.jpg"},{"id":"https://example.com/page5/","url":"https://example.com/page5/","title":"PAGE 5","content_html":"Content of Page 5","date_published":"Thu, 21 Jun 1979 23:45:00 GMT","date_modified":"Thu, 21 Jun 1979 23:45:00 GMT"}]}',
    data: {
      basename: "feed",
      content: '{"version":"https://jsonfeed.org/version/1","title":"My RSS Feed","home_page_url":"https://example.com/","feed_url":"https://example.com/feed.json","description":"","items":[{"id":"https://example.com/pages/new-name/page7/","url":"https://example.com/pages/new-name/page7/","content_html":"Content of Page 7","date_published":"Sun, 02 Jan 2022 00:00:00 GMT","date_modified":"Sun, 02 Jan 2022 00:00:00 GMT"},{"id":"https://example.com/pages/page6/","url":"https://example.com/pages/page6/","title":"PAGE 6","content_html":"<p>Content of Page 6</p>\\\\n","date_published":"Sat, 01 Jan 2022 00:00:00 GMT","date_modified":"Sat, 01 Jan 2022 00:00:00 GMT"},{"id":"https://example.com/pages/page4/","url":"https://example.com/pages/page4/","title":"PAGE 4","content_html":"Content of Page 4 in Overrided site name, from the file /pages/2021-01-02-18-32_page4.page.ts","date_published":"Sat, 02 Jan 2021 18:32:00 GMT","date_modified":"Sat, 02 Jan 2021 18:32:00 GMT"},{"id":"https://example.com/overrided-page2/","url":"https://example.com/overrided-page2/","title":"PAGE 2","content_html":"Content of Page 2","date_published":"Sun, 21 Jun 2020 00:00:00 GMT","date_modified":"Sun, 21 Jun 2020 00:00:00 GMT","image":"https://example.com/images/avatar.jpg"},{"id":"https://example.com/page5/","url":"https://example.com/page5/","title":"PAGE 5","content_html":"Content of Page 5","date_published":"Thu, 21 Jun 1979 23:45:00 GMT","date_modified":"Thu, 21 Jun 1979 23:45:00 GMT"}]}',
      page: [
        "src",
        "data",
        "asset",
      ],
      url: "feed.json",
    },
    src: {
      asset: true,
      ext: "",
      path: "",
      remote: undefined,
    },
  },
  {
    content: '<?xml version="1.0" encoding="UTF-8"?>
<rss xmlns:content="http://purl.org/rss/1.0/modules/content/" xmlns:wfw="http://wellformedweb.org/CommentAPI/" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:atom="http://www.w3.org/2005/Atom" xmlns:sy="http://purl.org/rss/1.0/modules/syndication/" xmlns:slash="http://purl.org/rss/1.0/modules/slash/" version="2.0">
  <channel>
    <title>My RSS Feed</title>
    <link>https://example.com/</link>
    <atom:link href="https://example.com/feed.rss" rel="self" type="application/rss+xml"/>
    <description></description>
    <lastBuildDate>Wed, 01 Jan 2020 00:00:00 GMT</lastBuildDate>
    <language>en</language>
    <generator>https://lume.land</generator>
    <item>
      <link>https://example.com/pages/new-name/page7/</link>
      <guid isPermaLink="false">https://example.com/pages/new-name/page7/</guid>
      <content:encoded><![CDATA[Content of Page 7]]></content:encoded>
      <pubDate>Sun, 02 Jan 2022 00:00:00 GMT</pubDate>
      <atom:updated>2022-01-02T00:00:00.000Z</atom:updated>
    </item>
    <item>
      <title>PAGE 6</title>
      <link>https://example.com/pages/page6/</link>
      <guid isPermaLink="false">https://example.com/pages/page6/</guid>
      <content:encoded>
        <![CDATA[<p>Content of Page 6</p>
        ]]>
      </content:encoded>
      <pubDate>Sat, 01 Jan 2022 00:00:00 GMT</pubDate>
      <atom:updated>2022-01-01T00:00:00.000Z</atom:updated>
    </item>
    <item>
      <title>PAGE 4</title>
      <link>https://example.com/pages/page4/</link>
      <guid isPermaLink="false">https://example.com/pages/page4/</guid>
      <content:encoded><![CDATA[Content of Page 4 in Overrided site name, from the file /pages/2021-01-02-18-32_page4.page.ts]]></content:encoded>
      <pubDate>Sat, 02 Jan 2021 18:32:00 GMT</pubDate>
      <atom:updated>2021-01-02T18:32:00.000Z</atom:updated>
    </item>
    <item>
      <title>PAGE 2</title>
      <link>https://example.com/overrided-page2/</link>
      <guid isPermaLink="false">https://example.com/overrided-page2/</guid>
      <content:encoded><![CDATA[Content of Page 2]]></content:encoded>
      <pubDate>Sun, 21 Jun 2020 00:00:00 GMT</pubDate>
      <atom:updated>2020-06-21T00:00:00.000Z</atom:updated>
      <meta property="og:image" content="https://example.com/images/avatar.jpg"/>
    </item>
    <item>
      <title>PAGE 5</title>
      <link>https://example.com/page5/</link>
      <guid isPermaLink="false">https://example.com/page5/</guid>
      <content:encoded><![CDATA[Content of Page 5]]></content:encoded>
      <pubDate>Thu, 21 Jun 1979 23:45:00 GMT</pubDate>
      <atom:updated>1979-06-21T23:45:00.000Z</atom:updated>
    </item>
  </channel>
</rss>',
    data: {
      basename: "feed",
      content: '<?xml version="1.0" encoding="UTF-8"?>
<rss xmlns:content="http://purl.org/rss/1.0/modules/content/" xmlns:wfw="http://wellformedweb.org/CommentAPI/" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:atom="http://www.w3.org/2005/Atom" xmlns:sy="http://purl.org/rss/1.0/modules/syndication/" xmlns:slash="http://purl.org/rss/1.0/modules/slash/" version="2.0">
  <channel>
    <title>My RSS Feed</title>
    <link>https://example.com/</link>
    <atom:link href="https://example.com/feed.rss" rel="self" type="application/rss+xml"/>
    <description></description>
    <lastBuildDate>Wed, 01 Jan 2020 00:00:00 GMT</lastBuildDate>
    <language>en</language>
    <generator>https://lume.land</generator>
    <item>
      <link>https://example.com/pages/new-name/page7/</link>
      <guid isPermaLink="false">https://example.com/pages/new-name/page7/</guid>
      <content:encoded><![CDATA[Content of Page 7]]></content:encoded>
      <pubDate>Sun, 02 Jan 2022 00:00:00 GMT</pubDate>
      <atom:updated>2022-01-02T00:00:00.000Z</atom:updated>
    </item>
    <item>
      <title>PAGE 6</title>
      <link>https://example.com/pages/page6/</link>
      <guid isPermaLink="false">https://example.com/pages/page6/</guid>
      <content:encoded>
        <![CDATA[<p>Content of Page 6</p>
        ]]>
      </content:encoded>
      <pubDate>Sat, 01 Jan 2022 00:00:00 GMT</pubDate>
      <atom:updated>2022-01-01T00:00:00.000Z</atom:updated>
    </item>
    <item>
      <title>PAGE 4</title>
      <link>https://example.com/pages/page4/</link>
      <guid isPermaLink="false">https://example.com/pages/page4/</guid>
      <content:encoded><![CDATA[Content of Page 4 in Overrided site name, from the file /pages/2021-01-02-18-32_page4.page.ts]]></content:encoded>
      <pubDate>Sat, 02 Jan 2021 18:32:00 GMT</pubDate>
      <atom:updated>2021-01-02T18:32:00.000Z</atom:updated>
    </item>
    <item>
      <title>PAGE 2</title>
      <link>https://example.com/overrided-page2/</link>
      <guid isPermaLink="false">https://example.com/overrided-page2/</guid>
      <content:encoded><![CDATA[Content of Page 2]]></content:encoded>
      <pubDate>Sun, 21 Jun 2020 00:00:00 GMT</pubDate>
      <atom:updated>2020-06-21T00:00:00.000Z</atom:updated>
      <meta property="og:image" content="https://example.com/images/avatar.jpg"/>
    </item>
    <item>
      <title>PAGE 5</title>
      <link>https://example.com/page5/</link>
      <guid isPermaLink="false">https://example.com/page5/</guid>
      <content:encoded><![CDATA[Content of Page 5]]></content:encoded>
      <pubDate>Thu, 21 Jun 1979 23:45:00 GMT</pubDate>
      <atom:updated>1979-06-21T23:45:00.000Z</atom:updated>
    </item>
  </channel>
</rss>',
      page: [
        "src",
        "data",
        "asset",
      ],
      url: "feed.rss",
    },
    src: {
      asset: true,
      ext: "",
      path: "",
      remote: undefined,
    },
  },
  {
    content: "<!DOCTYPE html>
<p>This page is exported to <code>/404.html</code>, not <code>/404/index.html</code></p>
",
    data: {
      basename: "404",
      children: "<p>This page is exported to <code>/404.html</code>, not <code>/404/index.html</code></p>
",
      content: "This page is exported to \`/404.html\`, not \`/404/index.html\`
",
      date: [],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Default site name",
      tags: "Array(0)",
      url: "/404.html",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/404",
      remote: undefined,
    },
  },
  {
    content: "<!DOCTYPE html>
Content of Page 5",
    data: {
      basename: "page5",
      children: "Content of Page 5",
      content: "Content of Page 5",
      date: [],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Default site name",
      tags: "Array(0)",
      title: "Page 5",
      url: "/page5/",
    },
    src: {
      asset: false,
      ext: ".yaml",
      path: "/page5",
      remote: undefined,
    },
  },
  {
    content: "<!DOCTYPE html>
Content of Page 2",
    data: {
      basename: "page2",
      children: "Content of Page 2",
      colors: "Array(3)",
      content: "Content of Page 2",
      date: [],
      documents: "Array(3)",
      drinks: [
        "alcoholic",
        "others",
      ],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
        "image",
      ],
      names: "Array(2)",
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Folder overrided site name",
      tags: "Array(1)",
      title: "Page 2",
      url: "/overrided-page2/",
    },
    src: {
      asset: false,
      ext: ".page.json",
      path: "/pages/2020-06-21_page2",
      remote: undefined,
    },
  },
  {
    content: "<!DOCTYPE html>
Content of Page 4 in Overrided site name, from the file /pages/2021-01-02-18-32_page4.page.ts",
    data: {
      basename: "page4",
      children: "Content of Page 4 in Overrided site name, from the file /pages/2021-01-02-18-32_page4.page.ts",
      colors: "Array(3)",
      content: "default",
      date: [],
      documents: "Array(3)",
      drinks: [
        "alcoholic",
        "others",
      ],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      names: "Array(2)",
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Overrided site name",
      tags: "Array(1)",
      title: "Page 4",
      url: "/pages/page4/",
    },
    src: {
      asset: false,
      ext: ".page.ts",
      path: "/pages/2021-01-02-18-32_page4",
      remote: undefined,
    },
  },
  {
    content: "<!DOCTYPE html>
<p>Content of Page 6</p>
",
    data: {
      basename: "page6",
      children: "<p>Content of Page 6</p>
",
      colors: "Array(3)",
      content: "Content of Page 6
",
      date: [],
      documents: "Array(3)",
      drinks: [
        "alcoholic",
        "others",
      ],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      names: "Array(2)",
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Folder overrided site name",
      tags: "Array(2)",
      title: "Page 6",
      url: "/pages/page6/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/pages/ghost/2021-12-29-page6",
      remote: undefined,
    },
  },
  {
    content: "Content of Page 3",
    data: {
      basename: "page3",
      colors: "Array(3)",
      content: "Content of Page 3",
      date: [],
      documents: "Array(3)",
      drinks: [
        "alcoholic",
        "others",
      ],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      names: "Array(2)",
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Folder overrided site name",
      tags: "Array(1)",
      title: "Page 3",
      url: "/page_3",
    },
    src: {
      asset: false,
      ext: ".page.js",
      path: "/pages/page3",
      remote: undefined,
    },
  },
  {
    content: "<!DOCTYPE html>
Content of Page 7",
    data: {
      basename: "page7",
      children: "Content of Page 7",
      colors: "Array(3)",
      content: "Content of Page 7",
      date: [],
      documents: "Array(3)",
      drinks: [
        "alcoholic",
        "others",
      ],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      names: "Array(2)",
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Folder overrided site name",
      tags: "Array(3)",
      url: "/pages/new-name/page7/",
    },
    src: {
      asset: false,
      ext: ".page.js",
      path: "/pages/subpage/page7",
      remote: undefined,
    },
  },
]
`;
